build-tools/packages/build-cli/src/commands/publish/tarballs.ts

```diff
@@ -0,0 +1,226 @@
+		}),
+	};
+
+	public async run(): Promise<void> {
+		const {
+			access,
+			dir,
+			dryRun,
+			orderFile,
+			retry,
+			tarball: orderFileIsTarballs,
+			publishArgs: rawPublishArgs,
+		} = this.flags;
+		const publishArgs: string[] = (rawPublishArgs ?? "").split(" ");
+		if (access !== undefined) {
+			publishArgs.push("--access", access);
+		}
+		if (dryRun) {
+			publishArgs.push("--dry-run");
+		}
+
+		const packageOrder = await readLines(orderFile).then((lines) =>
+			// filter out empty lines
+			lines.filter((line) => line !== undefined && line !== ""),
+		);
+
+		const tarballs = await globby(["*.tgz"], { cwd: dir, absolute: true });
+		const tarballMetadata = new Map<string, TarballMetadata>();
+
+		const mapPromises: Promise<void>[] = [];
+		for (const tarballPath of tarballs) {
+			mapPromises.push(
+				extractPackageJsonFromTarball(tarballPath).then((json) => {
+					const tarballName = getTarballName(json);
+					tarballMetadata.set(tarballName, {
+						name: json.name,
+						version: json.version,
+						filePath: tarballPath,
+						fileName: path.basename(tarballPath),
+					});
+				}),
+			);
+		}
+		await Promise.all(mapPromises);
+
+		for (const entry of packageOrder) {
+			const lookupEntry = orderFileIsTarballs ? entry : getTarballName(entry);
+			const toPublish = tarballMetadata.get(lookupEntry);
+			if (toPublish === undefined) {
+				this.error(`No tarball found matching '${entry}'`, { exit: 1 });
+			}
+
+			let tryCount = 0;
+			let status: PublishStatus;
+
+			do {
+				this.info(`Publishing ${toPublish.fileName}, attempt ${tryCount + 1}`);
+				// eslint-disable-next-line no-await-in-loop -- This should execute serially
+				status = await publishTarball(toPublish, this.logger, publishArgs);
+				tryCount++;
+			} while (status === "Error" && tryCount <= retry);
+
+			switch (status) {
+				case "AlreadyPublished": {
+					this.info(`Already published ${toPublish.fileName}, skipping`);
+					break;
+				}
+
+				case "SuccessfullyPublished": {
+					this.info(`Published ${toPublish.fileName}`);
+					break;
+				}
+
+				case "Error": {
+					this.error(`Fatal error publishing ${toPublish.fileName}, attempts: ${tryCount}`);
+				}
+
+				default: {
+					this.error(`Unexpected publish status: ${status}`, { exit: 1 });
+				}
+			}
+		}
+	}
+}
+
+/**
+ * Reads package.json from a gzipped tarball.
+ *
+ * Implementation from
+ * https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/3729bc2a3ca2ef7dda5c22fef81f89e1abe5dacf/packages/core/src/createPackage.ts#L296
+ */
+async function extractPackageJsonFromTarball(
+	tarballPath: string,
+): Promise<Readonly<PackageJson>> {
+	const tarball = new Uint8Array(await readFile(tarballPath));
+
+	// Use streaming API to work around https://github.com/101arrowz/fflate/issues/207
+	let unzipped: Uint8Array;
+	// eslint-disable-next-line no-return-assign
+	new Gunzip((chunk: Uint8Array) => (unzipped = chunk)).push(tarball, /* final */ true);
```

Comment:
You could put the new Gunzip in a local { }.
```suggestion
	{
		const gunzip = new Gunzip((chunk: Uint8Array) => (unzipped = chunk));
		gunzip.push(tarball, /* final */ true);
	}
```

Reply:
This doesn't address the lint error, which is complaining about the `unzipped = chunk` line.

Reply:
Yep - I was confused by the complicated line.
Still an improvement :)
I guess Gunzip callback requires a return value. The alternate is to put it is a block and also have a return, but I wouldn't change it. The comment is good.